### PR TITLE
[1/2] Add Kernel support for Cutlass based Fused FP4 MoE

### PR DESCRIPTION
**Pull request by pavanimajety**
This kernel adds support for NVFP4 MoE kernels. 

Currently measured perf: 
```
[--------------------------------------------------------------------------------------------- FP4 MOE vs FP8 Triton ---------------------------------------------------------------------------------------------]
                                                                                                                       |  triton_moe  |  triton_moe_cuda_graphs  |  cutlass_moe_fp4  |  cutlass_moe_fp4_cuda_graphs
1 threads: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((4, 2048, 7168))     |     10.7     |            9.9           |        12.3       |              10.5
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((8, 2048, 7168))     |     17.5     |           16.7           |        16.4       |              14.5
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((16, 2048, 7168))    |     29.5     |           28.8           |        23.4       |              21.5
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((32, 2048, 7168))    |     45.8     |           45.1           |        32.8       |              31.2
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((64, 2048, 7168))    |     60.9     |           60.0           |        41.5       |              40.3
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((128, 2048, 7168))   |     69.5     |           68.8           |        47.8       |              45.3
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((256, 2048, 7168))   |     72.3     |           71.4           |        50.9       |              49.9
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((512, 2048, 7168))   |     80.9     |           80.1           |        59.6       |              57.8
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((1024, 2048, 7168))  |     93.7     |           92.8           |        75.9       |              74.7
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((2048, 2048, 7168))  |    142.3     |          142.0           |       109.2       |             107.6

Times are in milliseconds (ms).
```

Follow up PR would include model support for `nvidia/DeepSeekR1-FP4`

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by Fridge003** [comment]
Hi @pavanimajety, please fix the lint by running `make format` under `sgl-kernel` folder

File: python/sglang/srt/layers/moe/cutlass_moe.py [hunk 1]
```diff
@@ -205,3 +204,153 @@ def cutlass_fused_experts(
...
+    assert 2 * half_k_w1 == k_w2, "Hidden size mismatch w2 and w1"
+    assert a.dtype in [torch.half, torch.bfloat16], "Invalid input dtype"
+    assert (
+        topk_weights.shape[0] == m and topk_ids.shape[0] == m
+    ), "topk must be provided for each row of a"
+
+    out_dtype = a.dtype
+    num_topk = topk_ids.shape[1]
+
+    expert_offsets = torch.empty((e + 1), dtype=torch.int32, device=device)
```
Comment (elfiegg) [inline python/sglang/srt/layers/moe/cutlass_moe.py:288]: For fp8 bs these tensors are initiated after weights are loaded and are passed into the class for perf reasons. I suggest for fp4 we can stick to that

File: python/sglang/srt/layers/moe/cutlass_moe.py [hunk 2]
```diff
@@ -205,3 +204,153 @@ def cutlass_fused_experts(
...
+    # Problem size:  (num_experts, (m,2n,k))
+    problem_sizes1 = torch.empty((e, 3), dtype=torch.int32, device=device)
+    # Problem size:  (num_experts, (m,n,k))
+    problem_sizes2 = torch.empty((e, 3), dtype=torch.int32, device=device)
```
Comment (elfiegg) [inline python/sglang/srt/layers/moe/cutlass_moe.py:316]: For fp8 bs these tensors are initiated after weights are loaded, and are passed into this class for perf reasons. I suggest for fp4 we can stick to that

File: python/sglang/srt/layers/moe/cutlass_moe.py [hunk 3]
```diff
@@ -205,3 +204,153 @@ def cutlass_fused_experts(
...
+
+    a_map = torch.empty((topk_ids.numel()), dtype=torch.int32, device=device)
+    c_map = torch.empty((topk_ids.numel()), dtype=torch.int32, device=device)
+
+    # problem shapes should have [m, n, k]
+    # Note that problem sizes are based on logical number of elements.
+    blockscale_offsets = torch.zeros(e + 1, dtype=torch.int32, device=device)
```
Comment (elfiegg) [inline python/sglang/srt/layers/moe/cutlass_moe.py]: This tensor is also static and requires buffer filling. consider also moving it outside of the class and initiating it after weights are loaded

File: sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+  torch::Tensor b_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor out_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor a_scales_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor b_scales_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor alpha_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor layout_sfa = torch::empty({num_experts, 5}, options_int);
+  torch::Tensor layout_sfb = torch::empty({num_experts, 5}, options_int);
+  torch::Tensor c_strides1 = torch::full({num_experts}, output.stride(0), options_int);
+  torch::Tensor a_strides1 = torch::full({num_experts}, a.stride(0) * 2, options_int);
+  torch::Tensor b_strides1 = torch::full({num_experts}, b.stride(1) * 2, options_int);
```
Comment (elfiegg) [inline sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu]: have you tested graph mode, would it work? these static tensor initiations can be moved outside of this kernel

File: sgl-kernel/csrc/moe/prepare_moe_input.cu [hunk 5]
```diff
@@ -126,3 +146,68 @@ void prepare_moe_input(
...
+  compute_arg_sorts<<<num_experts, num_threads, 0, stream>>>(
+      static_cast<const int32_t*>(topk_ids.data_ptr()),
+      static_cast<int32_t*>(input_permutation.data_ptr()),
+      static_cast<int32_t*>(output_permutation.data_ptr()),
+      static_cast<int32_t*>(atomic_buffer.data_ptr()),
+      topk_ids.numel(),
+      topk_ids.size(1));
+}
+
+void prepare_moe_input_v2(
```
Comment (elfiegg) [inline sgl-kernel/csrc/moe/prepare_moe_input.cu]: is it better to merge this with prepare_moe_input and set a flag for passing in blockscale_offset (or simply flag it with fp4)? the interface looks mostly identical and it'd probably be confusing to name it v1, v2 while the intention is blockscaled fp4 and fp8, whose actual difference is layout/extra scaling factor?

**Review by elfiegg — COMMENTED** [review]
Thanks Pavani for the work! I left some comments and realized it might require some structure changes and hopefully they make sense to you.

File: sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu (hunk 4 above)
Comment (pavanimajety) [reply sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu]: yes, this has been tested with graph mode.

File: python/sglang/srt/layers/moe/cutlass_moe.py (hunk 3 above)
Comment (pavanimajety) [reply python/sglang/srt/layers/moe/cutlass_moe.py]: Changed the logic, so initializing only empty which should have no overhead

File: sgl-kernel/csrc/moe/prepare_moe_input.cu (hunk 5 above)
Comment (pavanimajety) [reply sgl-kernel/csrc/moe/prepare_moe_input.cu]: Done, passing it as an optional argument

File: python/sglang/srt/layers/moe/cutlass_moe.py (hunk 2 above)
Comment (pavanimajety) [reply python/sglang/srt/layers/moe/cutlass_moe.py:316]: Modified it- The initialization happens in the test now and for the checkpoint it will happen in the post weight loading phase(part of second PR)

**Comment by pavanimajety** [comment]
@zhyncs @elfiegg I addressed the feedback. The failing tests seem unrelated to the PR. Please let me know if you have more suggestions or if it is ready for merge. Thanks!

File: sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+  auto options_int = torch::TensorOptions().dtype(torch::kInt64).device(a.device());
+
+  torch::Tensor a_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor b_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor out_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor a_scales_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor b_scales_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor alpha_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor layout_sfa = torch::empty({num_experts, 5}, options_int);
+  torch::Tensor layout_sfb = torch::empty({num_experts, 5}, options_int);
```
Comment (elfiegg) [inline sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu:288]: Can we try moving all these static tensors out of the kernel?

Comment (zhyncs) [reply sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu:288]: agree

Comment (pavanimajety) [reply sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu:288]: These don't actually have any overhead though. It's just allocates memory during capture and reuses it during replay. It provides for a cleaner interface to have these inside the kernel.

Comment (pavanimajety) [reply sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu:288]: They also go out of scope after the kernel finishes running, which also makes it slightly better I think.

Comment (elfiegg) [reply sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu:288]: if torch fails to cache it - in eager mode these will be called again and again though? also the two comments seem contradictory to me - If they go out of scope, that means you need to reallocate them the next time you invoke the kernel, which doesn't seem ideal

**Comment by pavanimajety** [comment]
```
[--------------------------------------------------------------------------- NVFP4 Blockscaled CUTLASS MOE vs FP8 Tensor Scaled Triton ---------------------------------------------------------------------------]
                                                                                                                       |  triton_moe  |  triton_moe_cuda_graphs  |  cutlass_moe_fp4  |  cutlass_moe_fp4_cuda_graphs
1 threads: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((4, 2048, 7168))     |      9.6     |            8.8           |         9.3       |               8.4           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((8, 2048, 7168))     |     15.6     |           15.2           |        13.3       |              11.9           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((16, 2048, 7168))    |     27.9     |           27.6           |        19.3       |              18.4           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((32, 2048, 7168))    |     43.0     |           42.7           |        28.1       |              27.3           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((64, 2048, 7168))    |     57.8     |           57.1           |        36.4       |              35.9           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((128, 2048, 7168))   |     66.3     |           65.9           |        41.0       |              41.3           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((256, 2048, 7168))   |     70.7     |           70.1           |        45.0       |              44.7           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((512, 2048, 7168))   |     78.5     |           77.8           |        51.0       |              50.8           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((1024, 2048, 7168))  |     89.8     |           89.2           |        64.2       |              63.3           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((2048, 2048, 7168))  |    133.1     |          132.5           |        91.0       |              90.5           

```
Latest perf: with torch::empty inside the cuda graph replay.
